External/FEXCore/include/FEXCore/Utils/File.h

```diff
@@ -0,0 +1,254 @@
+#endif
+      }
+
+      /**
+       * @brief Gets a File object that points to stdout
+       */
+      static File GetStdOUT() {
+#ifndef _WIN32
+        return File(STDOUT_FILENO, false);
+#else
+        return File(GetStdHandle(STD_OUTPUT_HANDLE), false);
+#endif
+      }
+
+      /**
+       * @brief Gets a File object that points to stderr
+       */
+      static File GetStdERR() {
+#ifndef _WIN32
+        return File(STDERR_FILENO, false);
+#else
+        return File(GetStdHandle(STD_ERROR_HANDLE), false);
+#endif
+      }
+
+      /**
+       * @brief Returns if the file handle is valid.
+       */
+      bool IsValid() const { return IsValidHandle; }
+
+      /**
+       * @brief Flush the file contents to the output file backing.
+       *
+       * @return True if the flush occured.
+       */
+      bool Flush() {
+#ifndef _WIN32
+        return fsync(Handle) == 0;
+#else
+        return FlushFileBuffers(Handle);
+#endif
+      }
+
+      /**
+       * @brief Seek the file pointer location.
+       *
+       * @param Distance The distance to travel.
+       * @param Op The operation from where to start the travel.
+       *
+       * @return The current file pointer location or -1.
+       */
+      ssize_t Seek(ssize_t Distance, SeekOp Op) {
+#ifndef _WIN32
+        return lseek(Handle, Distance, TranslateSeek(Op));
+#else
+        LARGE_INTEGER NewDistance {
+          .QuadPart = Distance
+        };
+        LARGE_INTEGER NewPointer;
+        auto Result = SetFilePointerEx(Handle, NewDistance, &NewPointer, TranslateSeek(Op));
+        if (Result) {
+          return NewPointer.QuadPart;
+        }
+        // Some error, match Linux side.
+        return -1;
+#endif
+      }
+
+    protected:
+
+      File(FileHandleType Handle, bool ShouldClose)
+        : ShouldClose {ShouldClose}
+        , IsValidHandle {true}
+        , Handle {Handle}
+      {}
+    private:
+      bool ShouldClose{};
+      bool IsValidHandle{};
+
+      FileHandleType Handle;
+#ifndef _WIN32
+      static constexpr int DEFAULT_USER_PERMS = S_IRWXU | S_IRWXG | S_IRWXO;
+
+      static uint32_t TranslateModes(FileModes Modes) {
+        uint32_t Mode{};
+        if ((Modes & FileModes::READ) == FileModes::READ)
+          Mode |= O_RDONLY;
+        if ((Modes & FileModes::WRITE) == FileModes::WRITE)
+          Mode |= O_WRONLY;
+        if ((Modes & FileModes::CREATE) == FileModes::CREATE)
+          Mode |= O_CREAT;
+        if ((Modes & FileModes::TRUNCATE) == FileModes::TRUNCATE)
+          Mode |= O_TRUNC;
+
+        // Always enable CLOEXEC so that the FD is closed on execve.
+        // FEXCore never wants to leak FDs across execve using this interface.
+        Mode |= O_CLOEXEC;
+        return Mode;
+      }
+
+      static uint32_t TranslateSeek(SeekOp Op) {
+        switch (Op) {
+          case SeekOp::BEGIN: return SEEK_SET;
+          case SeekOp::CURRENT: return SEEK_CUR;
+          case SeekOp::END: return SEEK_END;
```

Comment:
should this have a default case marked as unreachable or something?

Reply:
Added default case for FEX_UNREACHABLE